### DISCUSSION TITLE
Old DAGs are still available on airflow UI if the dag directory is changed

### DISCUSSION THREAD

**Discussion by MishraIshita** [apache/airflow#36375] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.7.3

### What happened?

We have seen an issue with airflow 2.7.3 where if the AIRFLOW__CORE__DAGS_FOLDER is changed and the dags have already been loaded from previous DAG_DIRECTORY, the old dags from previous dag directory still keep on showing on airflow UI.

### What you think should happen instead?

The DAGs present only in the current DAG Directory should be available on airflow UI

### How to reproduce

1. Create 2.7.3 airflow instance
2. Push a "example_1" dag in the default dags directory /usr/local/airflow/dags and let the airflow load it on airflow U
3. Change the AIRFLOW__CORE__DAGS_FOLDER  Variable to /usr/local/airflow/dags_mount, this already has example_2 dags.
4. This will load the  new DAG example_2  on the AIrflow UI but the example_1 dag will still keep on showing on UI

### Operating System

debian

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
I think you are wrong or you have not waited long enough (i.e. until scheduler processes the new directory and updates dag status - the dags that disappear should be moved to inactive). Converted it into discussion.

If you still observe it, showing your DAGs before and atfter and screenshots (after sufficient time resulting from airflow configuration) are needed to show some evidences. Also explanation on how you push the dags and how they are synchronized would be useful (including explanation on how you do it for all the components of Airflow - particularly scheduler)

**Comment by MishraIshita** [comment]
Directory Structure:
<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/0bc4a70e-deb5-493e-bce5-14d00573b622">
When  Dags folder is set to `/opt/airflow/dags`, we see just one dag `dags_directory.py` is available on the airflow UI.
<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/8c6c1b26-84b2-4fa1-bf99-7a9c31c44536">
<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/7b517405-a20e-431b-bdcf-b170c7e57e60">
On changing the Dags folder to `/opt/airflow/dags_mount`, the dag  - `dags_mount_directory.py`- from `dags_mount` directory is loaded on AIrflow UI but the dag from previous dag folder - `dags` is still available. Hence, two dags are populated on airflow UI
`dags_directory.py`
`dags_mount_directory.py`
<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/2bc2196a-fe86-46bb-9ba6-e0ef81d34ab7">
<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/520875a5-abde-4ea9-bbfe-2df218d9e1db">


<img width="452" alt="image" src="https://github.com/apache/airflow/assets/44608916/79c7b4a4-52b9-4179-aa5a-fefd2387569b">

**Comment by MishraIshita** [comment]
Hi @potiuk 
Please find the above attached evidence for the bug.

**Reply by potiuk** [reply]
You should post logs of scheduler/dag file processor showing what happens - processing dags etc. Also the configuration you show is that of webserver not scheduler. my guess is that you have either one or more schedulers running in the background scanning the directory, or maybe you do not run them continuously (only after some time dags will be removed if they are missing). Or you have muttiple dags defined in one place. It's also possible that your DAG fileis importing the other dag file and adds it to globals.

Looking at your processes and scheduler/dag processor logs should help you wilt investgating it.

**Reply by MishraIshita** [reply]
Hi
I am attaching the scheduler and dag file processing the logs. The dags are in two different directories, I have added the directory structure in above comment. The issue is occurring because the dag file processor is not marking the dag in previous dag directory "is_active"=false.

[dag_processor_manager.log](https://github.com/apache/airflow/files/13861175/dag_processor_manager.log)

[scheduler_logs.log](https://github.com/apache/airflow/files/13861182/scheduler_logs.log)


I have also tested the 2.7.0  airflow version and didn't find this bug but seeing this bug starting 2.7.1 airflow version. 

Specifically all the above testing was done on 2.7.3 airflow version.

I believe it has arisen due to following code change done in 2.7.1, when I revert it back to original query in 2.7.0, I am able to fix the issue.
https://github.com/apache/airflow/blob/1431535040f8d74f8b8e26c6d356e134ade24811/airflow/models/dag.py#L3711-L3719

**Reply by potiuk** [reply]
Hm @pierrejeambrun , it does look like https://github.com/apache/airflow/pull/33357 caused this side effect.

**Reply by pierrejeambrun** [reply]
Looks like it yes. The goal was to isolate dagfileprocessor to not wrongly interfere with each other and act ousite of their directories, but in this particular case, we want to allow it so they can deactivate other dags from previous directories🤔

**Reply by MishraIshita** [reply]
Hi @pierrejeambrun  and @potiuk , I will be happy to work on fixing this bug. Please let me know the process for the same.

**Comment by MishraIshita** [comment]
Hi @potiuk 
Please find the attached logs

**Comment by MishraIshita** [comment]
Hi @pierrejeambrun / @potiuk  Does reverting back the change works here?

**Reply by pierrejeambrun** [reply]
If we revert it, the original bug will be back. Meaning that if we have multiple DagFileProcessors running on different dag directories, they will keep on deactivating each others dag.

I would say that reverting is not a good idea, but we need to think of something that would solve both use cases.
